pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go

```diff
@@ -0,0 +1,438 @@
+		// and the old discarded. See type-level comment for more details.
+		internal spanconfig.Store
+		handlers []handler
+	}
+
+	lastFrontierTS hlc.Timestamp // used to assert monotonicity across subscription attempts
+}
+
+var _ spanconfig.KVSubscriber = &KVSubscriber{}
+
+// spanConfigurationsTableRowSize is an estimate of the size of a single row in
+// the system.span_configurations table (size of start/end key, and size of a
+// marshaled span config proto). The value used here was pulled out of thin air
+// -- it only serves to coarsely limit how large the KVSubscriber's underlying
+// rangefeed buffer can get.
+const spanConfigurationsTableRowSize = 5 << 10 // 5 KB
+
+// New instantiates a KVSubscriber.
+func New(
+	stopper *stop.Stopper,
+	db *kv.DB,
+	clock *hlc.Clock,
+	rangeFeedFactory *rangefeed.Factory,
+	spanConfigurationsTableID uint32,
+	bufferMemLimit int64,
+	fallback roachpb.SpanConfig,
+	knobs *spanconfig.TestingKnobs,
+) *KVSubscriber {
+	spanConfigTableStart := keys.SystemSQLCodec.IndexPrefix(
+		spanConfigurationsTableID,
+		keys.SpanConfigurationsTablePrimaryKeyIndexID,
+	)
+	spanConfigTableSpan := roachpb.Span{
+		Key:    spanConfigTableStart,
+		EndKey: spanConfigTableStart.PrefixEnd(),
+	}
+	spanConfigStore := spanconfigstore.New(fallback)
+	if knobs == nil {
+		knobs = &spanconfig.TestingKnobs{}
+	}
+	s := &KVSubscriber{
+		stopper:             stopper,
+		db:                  db,
+		clock:               clock,
+		bufferMemLimit:      bufferMemLimit,
+		rangefeedFactory:    rangeFeedFactory,
+		spanConfigTableSpan: spanConfigTableSpan,
+		fallback:            fallback,
+		knobs:               knobs,
+		decoder:             newSpanConfigDecoder(),
+	}
+	s.mu.internal = spanConfigStore
+	return s
+}
+
+// Subscribe establishes a subscription (internally: rangefeed) over the global
+// store of span configs. It fires off an async task to do so, re-establishing
+// internally when retryable errors[1] occur and stopping only when the surround
+// stopper is quiescing or the context canceled. All installed handlers are
+// invoked in the single async task thread.
+//
+// [1]: It's possible for retryable errors to occur internally, at which point
+//      we tear down the existing subscription and re-establish another. When
+//      unsubscribed, the exposed spanconfig.StoreReader continues to be
+//      readable (though no longer incrementally maintained -- the view gets
+//      progressively staler overtime). Existing handlers are kept intact and
+//      notified when the subscription is re-established. After re-subscribing,
+//      the exported StoreReader will be up-to-date and continue to be
+//      incrementally maintained.
+func (s *KVSubscriber) Subscribe(ctx context.Context) error {
+	return s.stopper.RunAsyncTask(ctx, "spanconfig-kvsubscriber", func(ctx context.Context) {
+		ctx, cancel := s.stopper.WithCancelOnQuiesce(ctx)
+		defer cancel()
+
+		for r := retry.Start(base.DefaultRetryOptions()); r.Next(); {
```

Comment:
I'm not so sure we want to just use the retry options without ever resetting. Consider a temporary outage. Consider calling `Reset` on the `Retry` if the thing ran for some amount of time.

Reply:
Do you mean something like: 

```go	
        log.Warningf(ctx, "spanconfig-kvsubscriber failed with %v, retrying...", err)
	if attemptNo > maxAttemptsBeforReset {
		attemptNo = 0
		r.Reset()
	}
	continue
}
```

The max backoff with the default options is 1 second, seems fine to not reset after a temporary outage? Especially if we'll pick it back up in a second.

Reply:
I was thinking something like:

```go
for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {
                        const aWhile = 5 * time.Minute // arbitrary but much longer than a retry
                        started := timeutil.Now()
			if err := s.run(ctx); err != nil {
				if errors.Is(err, context.Canceled) {
					return // we're done here
				}
				ranFor := timeutil.Since(start)
				if ranFor > aWhile {
				    r.Reset()
				}
				log.Warningf(ctx, "spanconfig-kvsubscriber failed with %v after %v, retrying...", err, ranFor)
				continue
			}
```